src/util.test.ts

```diff
@@ -227,4 +228,154 @@ describe('util', () => {
       expect(result).toEqual([join('/home', 'grid/send-email/provider.suma')]);
     });
   });
+
+  describe('gitDiff', () => {
+    let execCommand: string;
+
+    type ExecSpy = (
+      command: string,
+      callback?: (
+        error: ExecException | null,
+        stdout: string,
+        stderr: string
+      ) => void
+    ) => ChildProcess;
+
+    describe('when commit hashes are valid', () => {
+      let execSpy: jest.SpyInstance;
+
+      beforeEach(() => {
+        execSpy = jest.spyOn(
+          cp,
+          'exec'
+        ) as unknown as jest.MockedFunction<ExecSpy>;
+        execSpy.mockImplementation(function (
+          command: string,
+          callback?: (
+            error: ExecException | null,
+            stdout: string,
+            stderr: string
+          ) => void
+        ): ChildProcess {
+          execCommand = command;
+          if (callback) {
+            callback(
+              null,
+              `grid/starwars/character-information/profile.supr\ngrid/starwars/character-information/maps/mock.suma\n`,
+              ''
+            );
+          }
+
+          return new ChildProcess();
+        });
+      });
+
+      it('should call exec with valid params', async () => {
+        await util.gitDiff(
+          '9b133c68b5d88b82dbaaa5046174a9a629f38a41',
+          '3d4622addaf8c894ab4ba5c012d7dba90f0f3a73'
+        );
+
+        expect(execCommand).toEqual(
+          'git diff --name-only 9b133c68b5d88b82dbaaa5046174a9a629f38a41..3d4622addaf8c894ab4ba5c012d7dba90f0f3a73'
+        );
+      });
+
+      it('should return files', async () => {
+        const result = await util.gitDiff(
+          '9b133c68b5d88b82dbaaa5046174a9a629f38a41',
+          '3d4622addaf8c894ab4ba5c012d7dba90f0f3a73'
+        );
+
+        expect(result).toEqual([
+          'grid/starwars/character-information/profile.supr',
+          'grid/starwars/character-information/maps/mock.suma',
+        ]);
+      });
+    });
+
+    describe("when commit hash does't exist", () => {
+      it('should throw error', async () => {
+        const execSpy = jest.spyOn(
+          cp,
+          'exec'
+        ) as unknown as jest.MockedFunction<ExecSpy>;
```

Comment:
`exec` function has multiple implementation with different arguments.

If I don't do casting, I get TypeScript error on the next line `execSpy.mockImplementation...`, complaining that the arguments of the exec function and the mock are not compatible.

Reply:
Oh, because your mock isn't implementing the entire API. Missing optional `options` argument. Not saying it is better, I just prefer type safety.

```ts
        const execSpy = jest.spyOn(
          cp,
          'exec'
        );
        execSpy.mockImplementation(function (
          command: string,
          options?: any,
          callback?: (
            error: ExecException | null,
            stdout: string,
            stderr: string
          ) => void
        ): ChildProcess {
          execCommand = command;

          if (typeof options === 'function') {
            callback = options;
          }

          if (callback) {
            callback(
              null,
              '',
              "fatal: ambiguous argument '9b133c68b5d88b82dbaaa5046174a9a629f38a41..wrong_commit_hash': unknown revision or path not in the working tree."
            );
          }

          return new ChildProcess();
        });
```

Reply:
🤦 definitely better